parser-typechecker/src/Unison/Util/Rope.hs

```diff
@@ -0,0 +1,197 @@
+{-# Language FunctionalDependencies #-}
+{-# Language DeriveFoldable #-}
+
+module Unison.Util.Rope 
+  (chunks, singleton, one, map, traverse, null, flatten, two, cons, uncons, snoc, unsnoc, 
+   Sized(..), Take(..), Drop(..), Reverse(..), Index(..), Rope,
+   ) 
+where
+
+import Prelude hiding (drop,take,reverse,map,traverse,null)
+import Data.Foldable (toList)
+import Control.DeepSeq (NFData(..))
+
+-- | Size-balanced binary tree of chunks. 
+-- The `Int` in the `Two` constructor is a cached size of that subtree.
+data Rope a
+  = Empty
+  | One {-# unpack #-} !a
+  | Two {-# unpack #-} !Int !(Rope a) !(Rope a)
+  deriving Foldable
+
+chunks :: Rope a -> [a]
+chunks = toList
+
+singleton, one :: Sized a => a -> Rope a
+one a | size a == 0 = Empty
+one a = One a
+singleton = one
+
+map :: Sized b => (a -> b) -> Rope a -> Rope b
+map f = \case
+  Empty -> Empty
+  One a -> one (f a)
+  Two _ l r -> two (map f l) (map f r)
+
+traverse :: (Applicative f, Sized b) => (a -> f b) -> Rope a -> f (Rope b)
+traverse f = \case
+  Empty -> pure Empty
+  One a -> one <$> f a
+  Two _ l r -> two <$> traverse f l <*> traverse f r
+
+-- typeclasses used for abstracting over the chunk type 
+class Sized a where size :: a -> Int
+class Take a  where take :: Int -> a -> a
+class Drop a  where drop :: Int -> a -> a
+class Index a elem where index :: Int -> a -> Maybe elem
+class Reverse a where reverse :: a -> a
+
+instance Sized a => Sized (Rope a) where
+  size = \case
+    Empty -> 0
+    One a -> size a
+    Two n _ _ -> n
+
+null :: Sized a => Rope a -> Bool
+null r = size r == 0
+
+flatten :: Monoid a => Rope a -> a
+flatten = mconcat . toList
+
+instance (Sized a, Semigroup a) => Semigroup (Rope a) where (<>) = mappend
+instance (Sized a, Semigroup a) => Monoid (Rope a) where
+  mempty = Empty
+  mappend r1 r2 = case (r1,r2) of
+    (Empty, k) -> k
+    (k, Empty) -> k
+    (One a0, k2) -> cons' (size a0) a0 k2
+    (k1, One aN) -> snoc' k1 (size aN) aN
+    (k1@(Two sz1 l1 r1), k2@(Two sz2 l2 r2))
+      | sz1 * 2 >= sz2 && sz2 * 2 >= sz1 -> Two (sz1 + sz2) k1 k2
+      | sz1 > sz2                        ->  l1 <> (r1 <> k2)
+      | otherwise                        -> (k1 <> l2) <> r2
+
+instance Reverse a => Reverse (Rope a) where
+  reverse = \case
+    One a -> One (reverse a)
+    Two sz l r -> Two sz (reverse r) (reverse l)
+    Empty -> Empty
+
+two :: Sized a => Rope a -> Rope a -> Rope a
+two r1 r2 = Two (size r1 + size r2) r1 r2
+
+threshold :: Int
+threshold = 32
```

Comment:
Any insight into where this came from? It seems quite small for things like ByteStrings and Text.

The [Rope for the Yi text editor](https://github.com/yi-editor/yi-rope/blob/master/src/Yi/Rope.hs#L186-L191) uses 1200 as their chunk size.

At any rate it'd be good to write down the rationale here so we can re-evaluate in the future (even if the current rationale is "it's a guess"

Reply:
It's a guess, but what you're going for is that a block copy of less than `threshold` is as fast or faster than creating a new subtree that points to the two blocks. [Scala's Vector](https://github.com/scala/scala/blob/v2.13.3/src/library/scala/collection/immutable/Vector.scala#L97-L99) uses 32 for this, as does Clojure's sequence type. 

If you're okay with worse performance on edits and want to optimize more for memory usage, then you can make the threshold higher. That makes sense for a text editor where the runtime just needs to be fast enough to be imperceptible.

Basically, by default, this data structure is meant to be as fast as possible for edits (for situations where it's being used as a text buffer), and if you want to later coalesce some chunks to improve memory usage, you could do that. (We could just expose a `Text.flatten` operation like we do for `Bytes`, and then you can achieve whatever chunk size you want - repeatedly `take 1200`, flatten, concat)

I'll add some color on this in comments though!

Reply:
Awesome, thanks for the context! I learned something new today ✅